### PR TITLE
Fix #9542 (allow files under node_modules to be included in the compilation).

### LINKED ISSUE #9542: TypeScript 2.0: index.js not generated for source files in node_modules (since 2.0.0-dev.20160701)
We use NPM packages for creating TypeScript libraries. This is great for TypeScript code which is used in multiple projects. We already use TypeScript 2.0 and since 20160701 there is a problem when compiling those sources. The `index.ts` file in those packages is compiled, but the resulting `index.js` is not emitted. Other dependencies in `index.ts` (as in the example below) are emitted as expected. Only `index.js` is missing. I think it should be possible to have NPM packages with just the TypeScript source files so you can use packages for shared code/libraries. This worked like a charm before 2.0.0-dev.20160701. Probably broken by #9459. Could be on purpose, but the behavior right now is not consistent. All sources are compiled and emitted, only the `index.js` files are missing which is strange and results in a headache ;-).

**TypeScript Version:**  2.0.0-dev.20160701 (2.0.0-dev.20160630 gives the right output).

**Code to reproduce**
Create the following structure:
```
node_modules\test-module\index.ts
node_modules\test-module\something.ts
test.ts
```
 With the following code:
```ts
// node_modules\test-module\something.ts
export const SomeString: string = "Hello world";
```

```ts
// node_modules\test-module\index.ts
export { SomeString } from "./something";
```

```ts
// test.ts
import { SomeString} from "test-module";

console.log(SomeString); // Should output `Hello world`
```

**Expected behavior:**
When compiling `test.ts` the following files should be emitted:
```
node_modules\test-module\index.js
node_modules\test-module\something.js
test.js
```

**Actual behavior:** 
The `index.js` file is missing and the output is:
```
node_modules\test-module\something.js
test.js
```
If 2.0.0-dev.20160630 is installed the output is as expected.

I understand that in some cases it is not desirable to compile the code in the `node_modules` folder. But in our case we emit the code to a `outDir` specified in `tsconfig.json`. As the complete source tree will be reflected to this `outDir` also the `node_modules` folder is created in the `outDir`. Just as expected. We can use something like webpack or Browserify to bundle the code. We think this should be possible. Just pure TypeScript code in the NPM package and no need to publish the compiled JS in there...

Maybe it is better to have a compiler option to disable compiling sources under `node_modules` so one can decide the desired behavior...

### PR DESCRIPTION
**Pull request by billti**
Fixes #9542 . It took me about an hour to fix the issue, and another 4 to figure out why the baselines were dropping files for the test changes 😩 (see comment on changes in `projectsRunner.ts`).

I'm still verifying the changes to a couple of existing baselines are correct, but wanted to get this out for review now.

### CODE REVIEW

**Comment by billti** [comment]
OK. The other baseline changes are as expected due to the rather non-obvious behavior described in the comment at https://github.com/Microsoft/TypeScript/blob/master/src/harness/compilerRunner.ts#L95 . So this should be good to go.

File: src/harness/projectsRunner.ts [hunk 1]
```diff
@@ -479,17 +479,27 @@ class ProjectRunner extends RunnerBase {
...
-                                    try {
-                                        return Harness.IO.readFile(getProjectOutputFolder(outputFile.fileName, compilerResult.moduleKind));
-                                    }
-                                    catch (e) {
-                                        return undefined;
-                                    }
-                                });
+                                // There may be multiple files with different baselines. Run all and report at the end, else
+                                // it stops copying the remaining emitted files from 'local/projectOutput' to 'local/project'.
+                                try {
```
Comment (mhegazy) [inline src/harness/projectsRunner.ts]: sorry about that. i had this fixed a while back in the transforms branch, but never ported it back.

File: tests/baselines/reference/nodeResolution6.js [hunk 2]
```diff
@@ -13,7 +13,5 @@ export declare var y;
 import y = require("a"); 
 
 
-//// [ref.js]
```
Comment (mhegazy) [inline tests/baselines/reference/nodeResolution6.js]: why is this removed?

**Comment by mhegazy** [comment]
👍

**Pull request merged**